homeassistant/components/calendar/__init__.py

```diff
@@ -7,179 +7,200 @@
+        calendar = None
+        try:
+            if hasattr(platform, 'async_get_handler'):
+                calendar = yield from platform.async_get_handler(
+                    hass, p_config, discovery_info)
+            elif hasattr(platform, 'get_handler'):
+                calendar = yield from hass.async_add_job(
+                    platform.get_handler, hass, p_config, discovery_info)
+            else:
+                raise HomeAssistantError("Invalid calendar platform.")
+
+            if calendar is None:
+                _LOGGER.error(
+                    "Failed to initialize calendar platform %s", p_type)
+                return
+
+        except Exception:
+            _LOGGER.exception("Error setting up platform %s", p_type)
+            return
 
-DEFAULT_CONF_TRACK_NEW = True
-DEFAULT_CONF_OFFSET = '!!'
+        calendars.extend(calendar)
+        yield from component.async_add_entities(calendar)
 
+    setup_tasks = [async_setup_platform(p_type, p_config) for p_type, p_config
+                   in config_per_platform(config, DOMAIN)]
 
-# pylint: disable=too-many-instance-attributes
-class CalendarEventDevice(Entity):
-    """A calendar event device."""
+    if setup_tasks:
+        yield from asyncio.wait(setup_tasks, loop=hass.loop)
 
-    # Classes overloading this must set data to an object
-    # with an update() method
-    data = None
+    @asyncio.coroutine
+    def async_platform_discovered(platform, info):
+        """Setup discovered platform."""
+        yield from async_setup_platform(platform, discovery_info=info)
 
-    # pylint: disable=too-many-arguments
-    def __init__(self, hass, data):
-        """Create the Calendar Event Device."""
-        self._name = data.get(CONF_NAME)
-        self.dev_id = data.get(CONF_DEVICE_ID)
-        self._offset = data.get(CONF_OFFSET, DEFAULT_CONF_OFFSET)
-        self.entity_id = generate_entity_id(
-            ENTITY_ID_FORMAT, self.dev_id, hass=hass)
+    discovery.async_listen_platform(hass, DOMAIN, async_platform_discovered)
 
-        self._cal_data = {
-            'all_day': False,
-            'offset_time': dt.dt.timedelta(),
-            'message': '',
-            'start': None,
-            'end': None,
-            'location': '',
-            'description': '',
-        }
+    return True
 
-        self.update()
 
-    def offset_reached(self):
-        """Have we reached the offset time specified in the event title."""
-        if self._cal_data['start'] is None or \
-           self._cal_data['offset_time'] == dt.dt.timedelta():
-            return False
+class Calendar(Entity):
+    """Entity for each calendar platform."""
+
+    def __init__(self, hass, name):
+        """Initialze calendar entity."""
+        self.hass = hass
+        self._name = name
 
-        return self._cal_data['start'] + self._cal_data['offset_time'] <= \
-            dt.now(self._cal_data['start'].tzinfo)
+        self._next_event = None
```

Comment:
Make this a property instead (`next_event`) that defaults to `None`. Child calendars should override it.